### ISSUE TITLE
Switch to Acorn

### ISSUE THREAD

**Issue by nzakas** [eslint/espree#200]
**tl;dr** I'm stopping work on bugs to convert Espree over to be powered by Acorn.

There's a lot of time and energy required to maintain a parser. It was never the intent of Espree to need to exist and since Esprima came back to life, the intent was to wait for Esprima to provide the hooks Espree needed to make the changes we'd need to continue functioning in the same way. That's taken a lot longer than we thought (still not on the roadmap), and so Espree has lingered on.

We'd like to get to a point where Espree is basically just a thin layer on top of an already-existing parser. As it turns out, Acorn already has the hooks we need to do that. So, in the interest of supporting an already-existing and well-supported parser, I'll start working on moving Espree to be built on top of Acorn. The intent is to provide the exact same interface that Espree currently has, effectively turning Espree into a translation layer between Acorn and Esprima outputs. 

Things to get working:

- [x] Comment attachment
- [x] arrowFunctions
- [x] binaryLiterals
- [x] blockBindings
- [x] classes
- [x] defaultParams
- [x] destructuring
- [x] experimentalObjectRestSpread
- [x] forOf
- [x] generators
- [x] globalReturn
- [x] jsx
- [x] modules
- [x] newTarget
- [x] objectLiteralComputedProperties
- [x] objectLiteralDuplicateProperties
- [x] objectLiteralShorthandMethods
- [x] objectLiteralShorthandProperties
- [x] octalLiterals
- [x] regexUFlag
- [x] regexYFlag
- [x] restParams
- [x] spread
- [x] ~~superInFunctions~~ (removed)
- [x] templateStrings
- [x] unicodeCodePointEscapes
- [x] All tests running as modules
- [ ] Ecma Features mix
- [x] Tokenize

**Comment by xjamundx** [comment]
:clap: Cool, let me know if you need any specific help!

**Comment by nzakas** [comment]
I'm going to try to dig into this this week and hopefully will have some tasks you can chip in on. In the meantime, take a look at the `acorn` branch to get an idea of what's involved.

**Comment by nzakas** [comment]
Working on this.

**Comment by nzakas** [comment]
As a side note, I'm going to remove `superInFunctions`. It doesn't actually make sense anymore since the spec changed such that `super` can't really be used in functions anymore.

**Comment by nzakas** [comment]
Getting close, waiting on a few compat fixes:

https://github.com/estree/estree/issues/98
https://github.com/marijnh/acorn/issues/326
https://github.com/RReverser/acorn-jsx/issues/28
https://github.com/marijnh/acorn/issues/325

**Comment by impinball** [comment]
Should this be noted in the README? It doesn't seem to be up to date in light of this bug.

**Comment by xjamundx** [comment]
whoa @nzakas you got that done super fast

**Comment by nzakas** [comment]
The readme will be updated when the work  complete.

**Comment by hzoo** [comment]
> effectively turning Espree into a translation layer between Acorn and Esprima outputs.

So we should be able to merge anything in https://github.com/babel/acorn-to-esprima so that babel-eslint/babel-jscs only deals with flow/es7 stuff?

**Comment by nzakas** [comment]
I'm not sure what you mean, can you explain some more?

**Comment by nzakas** [comment]
@xjamundx I had done some early work a while back, and @rreverser did a bunch as well. I just started from where we left off and fixed failing tests. :)

**Comment by hzoo** [comment]
I was thinking we could use espree in babel-eslint if we could move the functionality there. Like for the template string issues you'll probably find https://github.com/babel/acorn-to-esprima/blob/master/src/convertTemplateType.js. 

https://github.com/eslint/espree/blob/b8576c6260401b52b392e0df70932dd0c4ccb9d7/espree.js#L351 is from https://github.com/babel/acorn-to-esprima/blob/master/src/toToken.js

I guess I thought espree could use acorn-to-esprima methods so it's a single place to make updates but I guess not?

**Comment by nzakas** [comment]
There's definitely some things that useful, but the overall approach we're taking is a bit different. Here's the latest: https://github.com/eslint/espree/blob/issue200/espree.js

In short, we're doing the conversion on-the-fly instead of after the fact, which allows us to more easily mimic what Espree is doing today in terms of `ecmaFeatures` errors.

**Comment by hzoo** [comment]
Ok cool I see

**Comment by nzakas** [comment]
Basically done with this. Now just waiting for the various issues/PRs on Acorn and Acorn-JSX to land to fix the remaining broken tests.

**Comment by nzakas** [comment]
Getting closer on this. A bunch of my changes have been merged into Acorn. There are a couple of edge cases I'm still working on, mostly having to do with template strings. Hopefully will have this done in the next couple of weeks.

**Comment by nzakas** [comment]
Even closer than I thought after the recent Acorn release. There are just a couple of failing tests left for `parse`. Then I need to look at `tokenize()`.

**Comment by nzakas** [comment]
Just waiting for one bug fix from Acorn: https://github.com/ternjs/acorn/pull/345

**Comment by nzakas** [comment]
Working through some perf issues. As it is, this is slower than existing Espree. I've been working on cutting the time down and I'm pretty close.

With Espree 2.x:

```
CPU Speed is 3093 with multiplier 7500000
Performance Run #1:  2542.365766ms
Performance Run #2:  2522.98052ms
Performance Run #3:  2545.898633ms
Performance Run #4:  2531.269895ms
Performance Run #5:  2561.971175ms
Performance budget exceeded: 2542.365766ms (limit: 2424.8302618816683ms)
```

With Espree 3.x:

```
CPU Speed is 3093 with multiplier 7500000
Performance Run #1:  2677.714532ms
Performance Run #2:  2671.42152ms
Performance Run #3:  2687.464265ms
Performance Run #4:  2730.368188ms
Performance Run #5:  2718.3307489999997ms
Performance budget exceeded: 2687.464265ms (limit: 2424.8302618816683ms)
```
